Source/WebCore/platform/network/curl/CurlMultipartHandle.cpp

```diff
@@ -34,309 +34,260 @@
+    return makeUnique<CurlMultipartHandle>(client, WTFMove(*boundary));
 }
 
-CurlMultipartHandle::CurlMultipartHandle(CurlMultipartHandleClient& client, const String& boundary)
+CurlMultipartHandle::CurlMultipartHandle(CurlMultipartHandleClient& client, String&& boundary)
     : m_client(client)
-    , m_boundary(boundary)
+    , m_boundary(WTFMove(boundary))
 {
-
 }
 
-void CurlMultipartHandle::didReceiveData(const SharedBuffer& buffer)
+void CurlMultipartHandle::didReceiveMessage(const SharedBuffer& buffer)
 {
-    if (m_state == State::End)
+    if (m_state == State::WaitingForTerminate || m_state == State::End || m_didCompleteMessage || m_hasError)
         return; // The handler is closed down so ignore everything.
 
     m_buffer.append(buffer.data(), buffer.size());
 
     while (processContent()) { }
 }
 
-void CurlMultipartHandle::didComplete()
+void CurlMultipartHandle::completeHeaderProcessing()
 {
-    // Process the leftover data.
+    ASSERT(m_state == State::WaitingForHeaderProcessing || m_state == State::End);
+
+    if (m_state == State::End || m_hasError)
+        return; // The handler is closed down so ignore everything.
+
+    m_state = State::InBody;
+
     while (processContent()) { }
+}
 
-    if (m_state != State::End) {
-        // It seems we are still not at the end of the processing.
-        // Push out the remaining data.
-        m_client.didReceiveDataFromMultipart(SharedBuffer::create(WTFMove(m_buffer)));
-        m_state = State::End;
-    }
+void CurlMultipartHandle::didCompleteMessage()
+{
+    if (m_state == State::End)
+        return; // The handler is closed down so ignore everything.
+
+    m_didCompleteMessage = true;
+
+    if (m_state == State::WaitingForTerminate)
+        m_state = State::Terminating;
 
-    m_buffer.clear();
+    while (processContent()) { }
 }
 
 bool CurlMultipartHandle::processContent()
 {
-/*
-    The allowed transitions between the states:
-         Check Boundary
-               |
-      /-- In Boundary <----\
-     |         |            |
-     |     In Header        |
-     |         |            |
-     |     In Content       |
-     |         |            |
-     |    End Boundary ----/
-     |         |
-      \-----> End
-
-*/
     switch (m_state) {
-    case State::CheckBoundary: {
-        if (m_buffer.size() < m_boundary.length()) {
-            // We don't have enough data, so just skip.
+    case State::FindBoundaryStart:
+        FALLTHROUGH;
+    case State::InBody: {
+        size_t matchPos = notFound;
+        size_t endPos = notFound;
+        bool hasCloseDelimiter = false;
+
+        auto result = findBoundary(matchPos, endPos, hasCloseDelimiter);
+        if (!result) {
+            m_hasError = true;
+            m_state = State::End;
             return false;
         }
 
-        // Check for the boundary string.
-        size_t boundaryStart;
-        size_t lastPartialMatch;
+        if (m_state == State::InBody)
```

Comment:
Is it OK just ignore the preceding data before the first boundary in State::FindBoundaryStart.

Reply:
How should I implement it? Can you show me the modified code?

Reply:
How about handling it as an syntax error?

~~~
        if (result.isSyntaxError || m_state == State::FindBoundaryStart && result.processed) {
            m_hasError = true;
            m_state = State::End;
            return false;
        }
~~~

Reply:
The RFC states the following, and it must be considered a normal case for data to enter before the first boundary. Therefore, I don't think this case should be treated as an error.

https://datatracker.ietf.org/doc/html/rfc2046#section-5.1.1
```
   There appears to be room for additional information prior to the
   first boundary delimiter line and following the final boundary
   delimiter line.  These areas should generally be left blank, and
   implementations must ignore anything that appears before the first
   boundary delimiter line or after the last one.
```

Reply:
Good to know. Thank you.